### PR TITLE
fix(chrome-relay): resilient reconnect, MV3 persistence, and navigation re-attach

### LINKED ISSUE #15099: Chrome extension relay: frequent disconnects require manual re-attach (fixable in background.js)
# GitHub Issue Draft: Chrome Extension Relay Disconnects

**Repo:** https://github.com/openclaw/openclaw
**File:** Post as a new issue

---

## Title

Chrome extension relay: frequent disconnects require manual re-attach (fixable in background.js)

## Labels

`bug`, `browser`, `extension`

### Problem

The Chrome extension relay (`background.js`) disconnects frequently during normal use. The user attaches a tab via the toolbar button, uses it successfully, then minutes later the agent gets "tab not found (no attached Chrome tabs for profile 'chrome')." The user must manually re-click the toolbar icon to reattach.

This happens mid-workflow — e.g., agent navigates to site A successfully, then a few minutes later tries to navigate to site B and the connection is gone.

### Environment

- OpenClaw 2026.2.9
- Brave Browser (Chromium-based, supports chrome.debugger)
- macOS 15 (Apple Silicon)
- Extension loaded unpacked via `openclaw browser extension install`

### Root Cause Analysis

After reviewing `background.js`, three issues cause the disconnects:

#### 1. WebSocket drop nukes all debugger sessions (lines ~105-120)

`onRelayClosed()` calls `chrome.debugger.detach()` on **every** attached tab and clears all state when the relay WebSocket drops. Any transient WebSocket hiccup (network blip, service worker restart) destroys all debugger sessions even though the debugger attachment itself was fine.

```js
// Current behavior — relay drop = detach everything
function onRelayClosed(reason) {
  relayWs = null
  // ...
  for (const tabId of tabs.keys()) {
    void chrome.debugger.detach({ tabId }).catch(() => {})  // ← unnecessary
  }
  tabs.clear()
}
```

**Fix:** Don't detach debugger on relay loss. Keep debugger attached, mark relay as disconnected, reconnect WS, then re-announce existing sessions.

#### 2. No WebSocket auto-reconnect

`ws.onclose` and `ws.onerror` only call `onRelayClosed()` — no reconnect attempt. The connection stays dead until the user manually clicks the toolbar icon.

**Fix:** Add exponential backoff reconnect in `onRelayClosed()`:

```js
function onRelayClosed(reason) {
  relayWs = null
  // DON'T detach debugger sessions
  // Schedule reconnect
  scheduleReconnect()
}

let reconnectAttempt = 0
function scheduleReconnect() {
  const delay = Math.min(1000 * 2 ** reconnectAttempt, 30000) + Math.random() * 1000
  reconnectAttempt++
  setTimeout(async () => {
    try {
      await ensureRelayConnection()
      reconnectAttempt = 0
      // Re-announce all still-attached tabs
      for (const [tabId, tab] of tabs.entries()) {
        if (tab.state === 'connected') {
          // Send attachedToTarget events for each tab
        }
      }
    } catch {
      scheduleReconnect()
    }
  }, delay)
}
```

#### 3. MV3 service worker state loss (lines ~19-24)

All tab/session state is in-memory Maps. When Chrome kills the MV3 service worker (~30s idle without activity), state is lost. The debugger may still be attached at the Chrome level, but the extension forgot about it.

Note: Since Chrome 118, active `chrome.debugger` sessions keep the worker alive, and since Chrome 116, WebSocket traffic resets the idle timer. But if both are briefly inactive, the worker can still die.

**Fix:** Persist minimal state (`tabId`, `sessionId`, `targetId`) to `chrome.storage.session`. On worker startup, rehydrate and validate with `chrome.tabs.get()`.

#### 4. No tab lifecycle cleanup (minor)

No `chrome.tabs.onRemoved` or `chrome.tabs.onReplaced` listeners. Closed/replaced tabs leave stale entries in the Maps, which can cause "No attached tab" errors.

**Fix:** Add listeners to clean up state when tabs close or get replaced (prerender swap).

### Impact

This is the biggest friction point when using the Chrome relay. Every disconnect requires physical user intervention (clicking the toolbar button), which breaks automated workflows and makes the relay unreliable for anything beyond single-shot use.

### Suggested Priority

The fixes are straightforward and contained within `background.js`. The most impactful single change is #1 (don't detach on WS drop) combined with #2 (auto-reconnect), which would eliminate most user-facing disconnects.

### References

- [Chrome debugger API - onDetach reasons](https://developer.chrome.com/docs/extensions/reference/api/debugger)
- [MV3 service worker lifecycle](https://developer.chrome.com/docs/extensions/develop/concepts/service-workers/lifecycle)
- [WebSockets in extension service workers](https://developer.chrome.com/docs/extensions/how-to/web-platform/websockets)

### PR DESCRIPTION
**Pull request by codexGW**
The Chrome extension relay loses connection after page navigation, sleep/wake cycles, or MV3 service worker restarts and never recovers. Users must manually re-click the toolbar icon after every disconnect.

## Changes

- **Keep debugger sessions alive across WebSocket disconnects** — `onRelayClosed()` no longer detaches debuggers; sessions are preserved with reconnecting badge state
- **Auto-reconnect with exponential backoff and jitter** — unlimited retries (1s → 30s cap), re-announces attached tabs on reconnect
- **Persist state to `chrome.storage.session`** — tab/session state survives MV3 worker restarts; rehydrates on every worker boot via top-level init promise
- **Re-attach debugger after cross-origin navigation** — `target_closed` detach triggers auto re-attach with graduated retry delays; guards against stale session overwrites and restricted URLs
- **Per-tab operation locks** — prevents double-attach races on concurrent toggle/attach operations
- **`chrome.alarms` keepalive** — MV3-correct wake mechanism (`periodInMinutes: 0.5`) alongside existing `setInterval` ping (20s)
- **Safe server-side stale WebSocket replacement** — removes 409 rejection; serializes via promise chain; nulls `extensionWs` before closing old socket so close handler is a no-op; guards on `message`/`close` handlers
- **Child-session detach forwarding** — emits `Target.detachedFromTarget` for child sessions before cleanup
- **Tab lifecycle listeners** — `onRemoved`, `onReplaced` (with child session migration), `onCompleted`
- **Badge refresh** — on navigation complete and tab switch (handles worker restart losing badge state)
- **Pending request timeouts** — 30s timeout prevents memory leaks from dropped relay messages

## Permissions added

- `webNavigation` — badge refresh after navigation
- `alarms` — MV3 keepalive

## Testing

Tested on macOS with Brave (Chromium) across: cross-origin navigation (Amazon → Reddit → YouTube → FAA → Google), WebSocket disconnect/reconnect, service worker restart, manual detach/reattach cycles, and Chrome restart (stale WS replacement).

Reviewed by Codex 5.3 (GPT-5.3) — head-to-head comparison with #15817.

Closes #15099
Relates to #1160

### CODE REVIEW

**Comment by codexGW** [comment]
**Update: merged best-of-breed improvements from #15817**

Added a second commit with 4 improvements inspired by @derrickburns' parallel PR, while keeping our original architecture (session-preserving reconnect, `whenReady` init gate, module-scope debugger listeners, unlimited retry with jitter, `chrome.storage.session`):

1. **Per-tab operation locks** — `tabOperationLocks` Set prevents double-attach races on toggle and attach paths
2. **`chrome.alarms` keepalive** — MV3-correct wake mechanism (`periodInMinutes: 0.5`, at Chrome's minimum) alongside our existing `setInterval` ping
3. **Safe server-side stale WS replacement** — Removes the 409 rejection. Serializes via promise chain, nulls `extensionWs` before closing old socket so close handler is a no-op. Guards on `message`/`close` handlers. No more gateway restart needed after Chrome restart.
4. **Child-session detach forwarding** — Emits `Target.detachedFromTarget` for child sessions before cleanup

Reviewed by Codex 5.3 (head-to-head comparison of both PRs).

**Comment by codexGW** [comment]
Updated with improvements adopted from #15817:

- **Seamless navigation re-attach**: On `target_closed`, we no longer send detach events to the relay. Instead, the tab stays in our maps with a "connecting" badge and re-attaches after a 500ms grace period. This makes cross-origin navigation seamless from the gateway's perspective.
- **Stale WebSocket close guard**: When a replaced WS fires its close event, the handler checks `ws !== relayWs` and no-ops if the connection was already replaced. Prevents the old socket's close handler from nuking the new connection.
- **Keepalive catch-path fix**: `ensureRelayConnection()` can throw without triggering `onRelayClosed` (e.g. preflight fetch fails). The alarm handler now explicitly schedules reconnect on failure.
- **Manual detach guard**: Checks `tabs.has(tabId)` before the grace-period re-attach fires, preventing re-attachment of a tab the user manually detached during the 500ms window.
- **Code cleanup**: Removed all dev markers, converted to upstream comment style, single clean squashed commit.

**Comment by heisenbugged** [comment]
i wonder why this was closed. this bug is basically making openclaw's browser unusable for me

**Comment by derrickburns** [comment]
Me too
